lib/ember-app.js

```diff
@@ -50,30 +50,41 @@ function EmberApp(options) {
   this._app = AppFactory['default']();
 }
 
-EmberApp.prototype.boot = function() {
-  return this._app.boot();
+EmberApp.prototype.boot = function(options) {
+  return this._app.boot().then(function(app) {
+    return app.buildInstance().boot(options);
+  });
 };
 
-EmberApp.prototype.visit = function(url) {
+EmberApp.prototype.visit = function(request, response, options) {
+  var path = request.path;
   var doc = new SimpleDOM.Document();
   var rootElement = doc.body;
   var options = { isBrowser: false, document: doc, rootElement: rootElement };
 
-  return this._app.visit(url, options).then(function(instance) {
-    var head;
-    if (doc.head) {
-      head = HTMLSerializer.serialize(doc.head);
-    }
-    try {
-      return {
-        url: instance.getURL(), // TODO: use this to determine whether to 200 or redirect
-        title: doc.title,
-        head: head,
-        body: HTMLSerializer.serialize(rootElement) // This matches the current code; but we probably want `serializeChildren` here
-      };
-    } finally {
-      instance.destroy();
-    }
+  return this.boot(options).then(function(instance) {
+    var cookiesService = instance.lookup('service:cookies');
+    cookiesService.setProperties({
+      request: request,
+      response: response
+    });
```

Comment:
This only runs after all the application's instance initializers have completed which doesn't really work of course as the cookie service won't be able then to read/write cookies then if used in an instance initializer.